apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/ClientPool.cs

```diff
@@ -24,46 +27,66 @@
+        private static readonly ConcurrentDictionary<ClientCredentialKey, CredentialClientPool> s_clientPoolByCredential =
+            new ConcurrentDictionary<ClientCredentialKey, CredentialClientPool>();
 
         public static async Task<SpannerClient> AcquireClientAsync(
             ITokenAccess credentials = null,
             ServiceEndpoint endpoint = null)
         {
-            var key = new ClientPoolKey(credentials, endpoint ?? SpannerClient.DefaultEndpoint);
-            var poolEntry = s_clientEntryPool.GetOrAdd(key, k => new ClientPoolEntry(key));
-            var result = await poolEntry.AcquireClientFromEntryAsync().ConfigureAwait(false);
-            Logger.LogPerformanceCounter("SpannerClient.Count", () => s_clientEntryPool.Count);
+            var key = new ClientCredentialKey(credentials, endpoint);
+            var poolEntry = s_clientPoolByCredential.GetOrAdd(key, k => new CredentialClientPool(k));
+            var result = await poolEntry.AcquireClientAsync().ConfigureAwait(false);
+            Logger.LogPerformanceCounter("SpannerClient.TotalCount", () => s_clientPoolByCredential.Count);
             return result;
         }
 
-        /// <summary>
-        /// Callers should be careful not to call AcquireClient while CloseAll is still running.
-        /// It is possible to obtain a stale client.
-        /// </summary>
-        /// <returns></returns>
-        public static async Task CloseAllAsync()
+        //ReSharper disable once UnusedMember.Global
+        //Returns the total of all reference counts.
+        //For test purposes only.
+        // poolContents will be filled with the current contents of the pool and may not be null.
+        internal static int GetPoolInfo(StringBuilder poolContents)
+        {
+            GaxPreconditions.CheckNotNull(poolContents, nameof(poolContents));
+            int referenceCountTotal = 0;
+            poolContents.AppendLine("ClientPool.Contents:");
+            int i = 0;
+            foreach (var kvp in s_clientPoolByCredential)
+            {
+                poolContents.AppendLine($"s_clientPoolByCredential({i}) Key:${kvp.Key}");
+                referenceCountTotal += kvp.Value.DumpCredentialPoolContents(poolContents);
+                i++;
+            }
+            return referenceCountTotal;
+        }
+
+        public static void ReleaseClient(SpannerClient spannerClient,
+            ITokenAccess credentials = null,
+            ServiceEndpoint endpoint = null)
         {
-            Logger.Debug(() => "Shutting down all gRPC channels.");
-            await SpannerClient.ShutdownDefaultChannelsAsync().ConfigureAwait(false);
-            s_clientEntryPool.Clear();
+            if (spannerClient != null)
+            {
+                var key = new ClientCredentialKey(credentials, endpoint);
+                var poolEntry = s_clientPoolByCredential.GetOrAdd(key, k => new CredentialClientPool(k));
```

Comment:
Do we definitely want GetOrAdd here? If there's no pool for that key, do we want to allocate one and then release the client? (The overall effect would just be to create an empty pool, yes?)